examples/sw/simple_system/pa_test/pa_test.c

```diff
@@ -0,0 +1,112 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "simple_system_common.h"
+
+#define PMP_NAPOT 0x18
+#define PMP_R     0x01
+#define PMP_W     0x02
+#define PMP_X     0x04
+
+#define MSTATUS_MPP_M (3 << 11)
+#define MSTATUS_MPP_U (0 << 11)
+
+
+typedef struct  {
+  uintptr_t pointer0;
+  uintptr_t pointer1;
+} fat_ptr_t;
+
+
+void setup_pakey() {
+  // Set the key for PA in CSR
+  uint32_t key = 0xdeadbeef;
+  asm volatile ("csrw 0x7c2, %[key]"
+                : : [key] "r" (key) :);
+}
+
+void setup_pmp() {
+  // 0x00000000 ~ 0x0FFFFFFF is accessible
+  uint32_t base_addr  = 0x0;
+  uint32_t size       = 1 << 30;
+  uint32_t napot_size = (size / 2) - 1;
+  uint32_t pmpaddr = (base_addr + napot_size) >> 2;
+  uint32_t pmpcfg = PMP_NAPOT | PMP_R | PMP_W | PMP_X;
+  asm volatile ("csrw pmpaddr0, %0;"
+                "csrw pmpcfg0, %1;"
+                : : "r" (pmpaddr), "r" (pmpcfg) :);
+}
+
+
+fat_ptr_t pac(fat_ptr_t ptr, uint32_t context) {
+  asm volatile (".insn r 0x0b, 0, 0, %[pointer1], %[pointer0], %[context];"
+                : [pointer0] "+r"  (ptr.pointer0), [pointer1] "=r" (ptr.pointer1)
+                : [context] "r" (context)
+  );
+  return ptr;
+}
+
+fat_ptr_t aut(fat_ptr_t ptr, uint32_t context) {
+  asm volatile (".insn r4 0x0b, 1, 0, %[pointer0], %[pointer0], %[pointer1], %[context];"
+                : [pointer0] "+r" (ptr.pointer0)
+                : [pointer1] "r" (ptr.pointer1), [context] "r" (context)
+  );
+  return ptr;
+}
+
+void print_fat_ptr(fat_ptr_t ptr) {
+  puts("pointer0: ");
+  puthex(ptr.pointer0);
+  putchar('\n');
+  puts("pointer1: ");
+  puthex(ptr.pointer1);
+  putchar('\n');
+}
+
+
+void test1() {
+  // test for pac and aut instruction
+  fat_ptr_t ptr = { .pointer0 = 0x1000, .pointer1 = 0x0 };
+  const uint32_t context = 0xffff;
+
+  fat_ptr_t pac_ptr = pac(ptr,     context);
+  print_fat_ptr(pac_ptr);
+  fat_ptr_t aut_ptr = aut(pac_ptr, context);
+  print_fat_ptr(aut_ptr);
+
+  if (ptr.pointer0 != aut_ptr.pointer0) { // Assert !
+    puts("Error\n");
+  }
+}
+
+void test2() {
+  // test whether pmp error occurs when accessing with unauthenticated pointer
+  const int data = 'A';
+  fat_ptr_t ptr = { .pointer0 = (uintptr_t)&data, .pointer1 = 0x0 };
+  const uint32_t context = 0xffff;
+
+  fat_ptr_t pac_ptr = pac(ptr, context);
+  *(int*)pac_ptr.pointer0 = 'B'; // <- Error occurs due to use of unauthenticated pointer
+}
+
+void test() {
```

Comment:
This function should have `__attribute__((interrupt("user")))` so that it correctly returns back to M-mode afterwards.

Reply:
By using `__attribute__((interrupt("user")))`,  we can save necessary registers at the beginning of this function and restore the registers at the end, and then call the `uret` instruction. 
However, ibex doesn't seem to currently support the `uret` instruction. I changed the test function as follows and executed it, but it seemed that the `uret` instruction could not be executed correctly.
```c
void test() {
    test1();
}
```

Reply:
@mmxsrup I'm wrong, you should not be using that attribute. My mental model of the privilege modes and how `uret` should work was wrong. 

I'm not sure what the final instruction of this function should be, but I do agree that it shouldn't be `uret`. Probably something like `wfi` is a more sensible choice, the important thing is to cause an m-mode interrupt so the core is left in m-mode.

Reply:
I think it is important to switch back to M-mode if the test succeeds normally. Perhaps it is good to raise an exception by `ecall` and switch back to M mode.

Reply:
Yeah, I think I see how `ecall` is the right thing.

Reply:
I think you should just end the simulation, by calling `sim_halt`. There's no point in trapping and going to the trap handler (an exception would be if `SIM_CTRL_BASE` wasn't accessible from user mode).